gh-63284: Add support for TLS-PSK (pre-shared key) to the ssl module

grantramsay opened the pull request:
Add support for TLS-PSK (pre-shared key) to the ssl module (plus documentation and unit-tests).

The referenced issue (https://github.com/python/cpython/issues/63284) is ~10 years old, but the information is still valid.

I've tested this by:

- connecting python TLS-PSK client/server (as in the unit-tests)
- connecting to openssl s_client/s_server from python:
  - `openssl s_server -accept localhost:12345 -psk deadbeef -cipher PSK -verify_return_error -nocert`
  - `openssl s_client -connect localhost:12345 -psk deadbeef -cipher PSK -verify_return_error`
  
Things I'm still uncertain on:

- ~~It's not yet working with TLS 1.3 (I'm currently looking into this).~~ [Fixed]
- ~~Do I need to call `PyErr_WriteUnraisable()` if there's a python exception during a C callback?~~ [Fixed]

This is my first python contribution.
I have probably got a few things wrong, but I'm happy to fix them up.

Thanks!

<!-- gh-issue-number: gh-63284 -->
* Issue: gh-63284
<!-- /gh-issue-number -->

grantramsay committed:
gh-63284: Add support for TLS-PSK (pre-shared key) to the ssl module

grantramsay committed:
Fix TLS-PSK for TLS 1.3

SSL_CTX_set_session_id_context() is a server-side only operation.
Using this on the client-side is causing authentication errors

arhadthedev committed:
Merge branch 'main' into fix-issue-63284

gpshead commented on Modules/_ssl.c:
```diff
@@ -3167,10 +3171,14 @@ _ssl__SSLContext_impl(PyTypeObject *type, int proto_version)
        usage for no cost at all. */
     SSL_CTX_set_mode(self->ctx, SSL_MODE_RELEASE_BUFFERS);
 
+    /* Setting the session id context is a server-side only operation.
```
Can you make a separate PR for this little logic change about only calling set_session_id_context for servers?  It seems like a bugfix that can stand on its own.

grantramsay replied:
Issue: https://github.com/python/cpython/issues/105293
PR: https://github.com/python/cpython/pull/105295

gpshead replied:
I'm going to leave this comment unresolved as a reminder that the other PR needs to be merged first.

gpshead commented on Modules/_ssl.c:
```diff
@@ -4619,6 +4627,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+    }
+
+    PyObject *hint_str = (hint != NULL) ?
+            PyUnicode_DecodeASCII(hint, strlen(hint), "strict") :
```
Why ASCII? I suspect "hint" should really be defined as PyBytes rather than PyUnicode.

gramsay0 replied:
I changed this to UTF-8 as specified in the RFC and added a test with non-ASCII chars.
https://www.rfc-editor.org/rfc/rfc4279

> 5.1.  PSK Identity Encoding
The PSK identity MUST be first converted to a character string, and then encoded to octets using UTF-8.

gpshead commented on Doc/library/ssl.rst:
```diff
@@ -1987,6 +1987,89 @@ to speed up repeated connections from the same clients.
+   In general, certificate based authentication should be preferred over this method.
+
+   The parameter ``callback`` is a callable object with the signature:
+   ``def callback(hint: str | None) -> tuple[str | None, bytes]``.
```
per comment in the code, hint should probably be `bytes | None`

gramsay0 replied:
Changed to UTF-8 so keeping this as a string

gpshead commented on Doc/library/ssl.rst:
```diff
@@ -1987,6 +1987,89 @@ to speed up repeated connections from the same clients.
+          return 'ClientId_1', psk_table[hint]
+      context.set_psk_client_callback(callback)
+
+   .. versionadded:: 3.12
```
The versionadded fields will need to be 3.13, we've missed the 3.12 feature freeze.

grantramsay committed:
Decode TLS-PSK identities as UTF-8 rather than ASCII

RFC4279 states these are UTF-8.
Add unit test using non-ASCII chars

grantramsay committed:
Change TLS-PSK version added from 3.12 to 3.13

The PR has missed the 3.12 merge window

gpshead committed:
ReSTify NEWS.

gpshead committed:
Merge branch 'main' into fix-issue-63284

gpshead commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+    }
+
+    PyObject *hint_str = (hint != NULL) ?
+            PyUnicode_DecodeUTF8(hint, strlen(hint), "strict") :
```
how safe is it to call `strlen(hint)`? this is raw data being supplied by OpenSSL, does it guarantee a reasonable length and null termination for us? (no hint length is passed, so i'm assuming so)

should the hint be treated as UTF-8?  Does TLS or OpenSSL require that?  It might make more sense to have code use PyBytes instead.  A failed UTF-8 decode is something the other end of the network connection could trigger.  same for the other callback regarding identity.

gramsay0 replied:
> does it guarantee a reasonable length and null termination for us?

Yes:

> The callback function is given the connection in parameter ssl, a NULL-terminated PSK identity hint sent by the server in parameter hint

The OpenSSL implementation limits `identity` and `identity_hint` string length to `#define PSK_MAX_IDENTITY_LEN 256`, and uses `strndup` to guarantee null termination.
(Oddly this doesn't seem to meet identity length specified in the RFC `opaque psk_identity<0..2^16-1>;`)

> should the hint be treated as UTF-8? Does TLS or OpenSSL require that?

The RFC requires `identity` and `identity_hint` to be encoded as UTF-8:

> 5.1. PSK Identity Encoding
The PSK identity MUST be first converted to a character string, and then encoded to octets using UTF-8.

Although, the OpenSSL implementation doesn't do any UTF-8 validation, it treats what it is given as null terminated bytes. So it would be possible to cause a python exception on the remote side by sending invalid UTF-8 (I assume that is what you would like to avoid).

I think a good approach would be for the C callbacks to just ignore/drop requests with invalid UTF-8. Although this would require a UTF-8 validation function in C that does not raise a python exception (not sure if that exists).

For now I will follow your advice and change these to PyBytes

gramsay0 replied:
> For now I will follow your advice and change these to PyBytes

@gpshead I've changed my mind on this.
I don't think it makes sense to use PyBytes to handle a UTF-8 string.
This could stay as UTF-8 and call `PyErr_Clear();` if the UTF-8 decode fails and reject the connection.

Could you please elaborate on your concerns if you think this should still be PyBytes?

gpshead replied:
your reasoning makes sense and I like your PyErr_Clear approach when the Decode fails.

gpshead commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+    PyGILState_Release(gstate);
+    return (unsigned int)psk_len_;
+
+error:
```
Your thought in the PR text sounds correct: doing

```c
if (PyErr_Occurred()) {
    PyErr_WriteUnraiseable(callback);
    PyErr_Clear();
}
```

makes sense here before releasing the GIL.  same in the other callback.

gpshead commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+        goto error;
+    }
+
+    PyObject *identity_str = (identity[0] != '\0') ?
```
coding defensively, include `identity != NULL || ...` in that condition.

gpshead commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+    }
+
+    PyObject *identity_str = (identity[0] != '\0') ?
+            PyUnicode_DecodeUTF8(identity, strlen(identity), "strict") :
```
why utf-8 rather than bytes?

gpshead commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+
+    const char *psk_;
+    Py_ssize_t psk_len_;
+    if (!PyArg_Parse(result, "y#", &psk_, &psk_len_)) {
```
rather than using `PyArg_Parse` here, use:

if (PyBytes_AsStringAndSize(&psk_, &psk_len_) < 0) {
}

it's a single value, there is no need to parse anything.

grantramsay committed:
Set python exceptions raised during C callbacks as unraisable

grantramsay committed:
Add NULL check for defensive coding

grantramsay committed:
Get single value using PyBytes_AsStringAndSize

There is no need to use PyArg_Parse in this simple case

gramsay0 commented on Modules/_ssl.c:
```diff
@@ -3116,6 +3118,8 @@ _ssl__SSLContext_impl(PyTypeObject *type, int proto_version)
     self->set_sni_cb = NULL;
     self->state = get_ssl_state(module);
+    self->psk_client_callback = NULL;
+    self->psk_server_callback = NULL;
```
Free callbacks when SSL context is deallocated

gramsay0 commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+    }
+    self->psk_server_callback = callback;
+    SSL_CTX_set_psk_server_callback(self->ctx, ssl_callback);
+    SSL_CTX_use_psk_identity_hint(self->ctx, identity_hint);
```
Check return value of `SSL_CTX_use_psk_identity_hint`

gramsay0 commented on Doc/library/ssl.rst:
```diff
@@ -1987,6 +1987,89 @@ to speed up repeated connections from the same clients.
          >>> ssl.create_default_context().verify_mode  # doctest: +SKIP
          <VerifyMode.CERT_REQUIRED: 2>
 
+.. method:: SSLContext.set_psk_client_callback(callback)
```
Add documentation on:

- the length limits of the `identity` and `hint` parameters
- how to reject a connection (return a PSK of zero length)

gramsay0 commented on Modules/_ssl.c:
```diff
@@ -4641,6 +4645,202 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+    }
+
+    if (self->psk_server_callback != NULL) {
+        Py_DECREF(self->psk_server_callback);
```
Make use of `Py_XINCREF` and `Py_XDECREF` for tidier NULL checking

gramsay0 reviewed (commented):
Leaving myself notes on things to tidy up

grantramsay committed:
Do not raise a decode exception if remote side sends invalid UTF-8 during TLS-PSK negotiation

In this case the remote side is breaking the standard (rfc4279).
Just drop the connection

grantramsay committed:
Free TLS-PSK callbacks when SSL context is deallocated

grantramsay committed:
Use Py_XINCREF and Py_XDECREF for tidier NULL checking

grantramsay committed:
Check return value of SSL_CTX_use_psk_identity_hint

grantramsay committed:
Update TLS-PSK documentation

Add length limits of identity and hint parameters.
Add how to reject a connection

grantramsay committed:
Regenerate news entry to update the timestamp

gpshead committed:
Merge branch 'main' into fix-issue-63284

gpshead committed:
Update example key strings in the doc.

grantramsay committed:
Update TLS-PSK documentation

Demonstrate usage in a safer way that rejects
unknown connection attempts

grantramsay committed:
Regenerate news entry to update the timestamp

gpshead commented on Modules/_ssl.c:
```diff
@@ -4662,6 +4668,220 @@ _ssl__SSLContext_get_ca_certs_impl(PySSLContext *self, int binary_form)
+        goto error;
+    }
+    memcpy(psk, psk_, psk_len_);
+    memcpy(identity, identity_, identity_len_);
```
i added a check that `identity_` was not NULL here as the "z#" above can return NULL but doesn't specify if `identity_len_` will be set to anything in that situation.

gpshead committed:
Code defensively around `z#` vs NULL and 0 length.

gpshead committed:
Merge branch 'fix-issue-63284' of https://github.com/grantramsay/cpython into fix-issue-63284
